ADBench/plot_graphs.py

```diff
@@ -188,25 +219,78 @@ def sorting_key_fun(v):
-    failed_x, failed_y = [], []
-    for handle in failed:
-        inf_inds = [i for i, y in enumerate(handle.get_ydata()) if y == float("inf")]
-        last_ind = inf_inds[0] - 1 if len(inf_inds) > 0 else -1
-        failed_x.append(handle.get_xdata()[last_ind])
-        failed_y.append(handle.get_ydata()[last_ind])
-        #handles += pyplot.plot(handle.get_xdata()[last_ind], handle.get_ydata()[last_ind], marker="o", color=(0, 0, 0), linestyle="None", label="Crashed/Terminated")
-    if len(failed_x) > 0:
-        handles += tuple(pyplot.plot(failed_x, failed_y, marker="o", color="k", linestyle="None", label=TERMINATED_LABEL))
-        labels += (TERMINATED_LABEL,)
+    violation_x, violation_y = [], []
+    was_violation = False
+    additional = []
+    violation_handle = None
+
+    # Plot results
+    for ((color, marker), (tool, n_vals, t_vals, violations)) in lines:
+        def vals_with_neighbours_and_violation():
+            # Checking neighbours by shifting t_vals requires that
+            # it is in the order of monotonic n_vals
+            assert n_vals == sorted(n_vals)
+            return zip(
+                n_vals,
+                t_vals,
+                # Whether the left neighbour is missing
+                [True] + [t_val == float("inf") for t_val in t_vals],
+                # Whether the right neighbour is missing
+                [t_val == float("inf") for t_val in t_vals[1:]] + [True],
+                violations)
+
+        label = utils.format_tool(tool)
+        all_terminated = all(t_val == float("inf") for t_val in t_vals)
+        # Append label in legend if all point values are infinite
+        if all_terminated:
+            label += ALL_TERMINATED_SUFFIX
+
+        labels.append(label)
+        handles += pyplot.plot(
+            n_vals,
+            t_vals,
+            marker=marker,
+            color=color,
+            label=label
+        )
+
+        together = list(vals_with_neighbours_and_violation())
+        additionals = [(n_val, t_val)
+                       for (n_val, t_val, missing_left, missing_right, violation)
+                       in together
+                       if t_val != float("inf")
+                       and missing_left
+                       and missing_right
+                       and violation]
+
+        # addint coordinates of additional markers
+        additional.append(([n_val for (n_val, _) in additionals],
+                           [t_val for (_, t_val) in additionals],
+                           color))
+
+        violation_x += [n_val for (n_val, _, _, _, violation)
+                        in together if violation]
+        violation_y += [t_val for (_, t_val, _, _, violation)
+                        in together if violation]
+        was_violation = \
+            was_violation \
+            or any(violation for (_, _, _, _, violation) in together)
+
+
+    handles += pyplot.plot(
```

Comment:
I think, handle for violation points need to be added only if there was any violation, so, in my opinion, it should be moved to "if" body below

Reply:
We could have it inside the `if` or outside.  The behaviour will be the same.  The question is, what leads to simpler code.  I think it's simpler to have it outside.  Then we don't have to think about whether it's run or not.  It is always run and always does the right thing, even if there are no violations.  What do you think?

Reply:
This make sense for plotly, because when it will convert figure it will look through handles and will make legend by them, not by labels. So, here the example of your code result in plotly:
![image](https://user-images.githubusercontent.com/40039815/65820420-4b882800-e231-11e9-9550-3d231e0b35c0.png)
AS you can see, there is not any violations but violation handle is present

Reply:
Ah that's interesting.  Thank you.  I will fix that.

Reply:
Done